switchboard/src/service.rs

```diff
@@ -0,0 +1,1872 @@
+
+pub mod herd;
+pub mod kanban;
+mod socket_connection;
+mod tag;
+
+#[derive(Debug, Error)]
+pub enum ServiceError {
+    #[error("Database error: {0}")]
+    Database(sqlx::Error),
+    #[error("Herd error: {0}")]
+    Herd(HerdError),
+    #[error("Reservation error: {0}")]
+    Reservation(ReservationError),
+    #[error("Kanban error: {0}")]
+    Kanban(KanbanError),
+    #[error("Job's tag configuration did not match any currently registered supervisors")]
+    SupervisorMatchError,
+    #[error("Job deleted between match and launch")]
+    NoLongerQueued,
+    #[error("No such job")]
+    NoSuchJob,
+    #[error("Supervisor already connected")]
+    SupervisorAlreadyConnected,
+}
+
+struct State {
+    kanban: Kanban,
+    herd: Herd,
+}
+pub struct Service {
+    state: Mutex<State>,
+    pool: PgPool,
+    service_config: ServiceConfig,
+}
+
+// -------------------------------------------------------------------------------------------------
+// === GROUPING: SUPERVISOR INITIALISATION / NORMALISATION
+// -------------------------------------------------------------------------------------------------
+
+impl Service {
+    // ---- PUBLIC ----
+
+    #[instrument(skip(pool, service_config))]
+    pub async fn new(
+        pool: PgPool,
+        service_config: ServiceConfig,
+    ) -> Result<Arc<Self>, ServiceError> {
+        let this = Arc::new(Self {
+            state: Mutex::new(State {
+                kanban: Kanban::new(),
+                herd: Herd::new(),
+            }),
+            pool,
+            service_config,
+        });
+
+        // TODO: make sure there are no cases of multiple jobs running on the same supervisor
+
+        // Load supervisor table from the database.
+        this.load_supervisors().await?;
+
+        // Now, recover the state from before the switchboard last exited
+
+        // First, load the set of 'running' jobs and separate it into jobs that have hit timeout and
+        // jobs that have not. Anything that has hit timeout, we will then try to restart.
+        // The simplest way to 'restart', in this case, is to insert a new rows in the database,
+        // which `normalize_queued_job_states()` and `load_job_queue_from_vec()` will then pick up
+        // and rebuild, so we handle active jobs before queued jobs.
+        let (active_jobs, maybe_restartable) = this.normalize_active_job_states().await?;
+        this.load_active_jobs_from_vec(active_jobs).await;
+        this.try_restart_expired_jobs(maybe_restartable).await?;
+
+        // Load the set of 'queued' jobs, prune out any that timed out in the queue, and load the
+        // rest in the kanban.
+        let queued_jobs = this.normalize_queued_job_states().await?;
+        this.load_job_queue_from_vec(queued_jobs).await?;
+
+        // Launch the match daemon
+        Arc::clone(&this).launch_match_daemon();
+
+        Ok(this)
+    }
+
+    // ---- INTERNALS ----
+
+    fn launch_match_daemon(self: Arc<Self>) {
```

Comment:
Why don't we have one match daemon per supervisor, where each then tries to claim jobs? The database serves as the arbiter for which supervisor can claim a job, in case of concurrent claims.

Reply:
That's also possible but the current approach was simpler, and also makes it easy if we want to impose e.g. progress guarantees or similar.